demo-project/demo-frontend/callback.js

```diff
@@ -1,26 +1,52 @@
 // get the id_token and refresh_token from the fragment part of the url
 // and store them in the session storage
-const params = new URLSearchParams(window.location.hash.slice(1));
-const idToken = params.get('id_token');
-const refreshToken = params.get('refresh_token');
-const state = params.get('state');
+const fragmentParams = new URLSearchParams(window.location.hash.slice(1));
+const idToken = fragmentParams.get('id_token');
+const refreshToken = fragmentParams.get('refresh_token');
+const state = fragmentParams.get('state');
+
+const params = new URLSearchParams(window.location.search);
+const authCode = params.get('code');
 
 if (idToken && refreshToken && state) {
-    const fn = async ()=>{
-        // use token to access protected route
-        const resp = await fetch('/protected', {
-            headers: {
-                Authorization: `Bearer ${idToken}`
-            }
-        });
-        const greeting = document.createElement('h2');
-        if (resp.status !== 200) {
-            greeting.innerText = `Error: ${resp.status} ${resp.statusText}`;
-            document.body.appendChild(greeting);
-            return;
-        }
-        greeting.innerText = await resp.text();
-        document.body.appendChild(greeting);
-    };
-    fn();    
+  const fn = async () => {
+    // use token to access protected route
+    const resp = await fetch('/protected', {
+      headers: {
+        Authorization: `Bearer ${idToken}`
+      }
+    });
+    const greeting = document.createElement('h2');
+    if (resp.status !== 200) {
+      greeting.innerText = `Error: ${resp.status} ${resp.statusText}`;
+      document.body.appendChild(greeting);
+      return;
+    }
+    greeting.innerText = await resp.text();
+    document.body.appendChild(greeting);
+  };
+  fn();
+}
+
+if (authCode) {
```

Comment:
This should really be an `else if` right?

Reply:
You could also return the common stuff in either block, and move everything at the end. Each block should basically return `resp`, as far as I can see.

Reply:
This demo only shows the happy path, so `if` or `else if` doesn't make a big difference.
About the second point, the code is async and not awaited for some reason, probably to not block the UI. So we can't run the code afterwards, I just make a function to avoid the duplication.

c3657fce9d60
827e05422e1c